### DISCUSSION TITLE
Why is PikeVM the slowest engine?

### DISCUSSION THREAD

**Discussion by zackradisic** [rust-lang/regex#1121] [Q&A] [answered]
In the docs it says the PikeVM regex engine _"tends to be a tool of last resort because it is also usually the slowest regex engine_".

I feel like executing bytecode should be faster than dealing with the big graph data structures associated with the other DFA-based approaches

Can anyone offer any insight on why this is the case?

**Comment by BurntSushi** [comment, marked as answer]
I think I [answered this on HackerNews](https://news.ycombinator.com/item?id=38127639):

> Before getting to your actual question, it might help to look at a regex benchmark that compares engines (perhaps JITs are not the fastest in all cases!): https://github.com/BurntSushi/rebar
>
> In particular, the `regex-lite` engine is strictly just the PikeVM without any frills. No prefilters or literal optimizations. No other engines. Just the PikeVM.
>
> As to your question, the PikeVM is, essentially, an NFA simulation. The PikeVM just refers to the layering of capture state on top of the NFA simulation. But you can peel back the capture state and you're still left with a slow NFA simulation. I mention this because you seem to compare the PikeVM with "big graph structures with NFAs/DFAs." But the PikeVM *is* using a big NFA graph structure.
>
> At a very high level, the time complexity of a Thompson NFA simulation and a DFA hints strongly at the answer to your question: searching with a Thompson NFA has worst case O(m\*n) time while a DFA has worst case O(n) time, where m is proportional to the size of the regex and n is proportional to the size of the haystack. That is, for each character of the haystack, the Thompson NFA is potentially doing up to `m` amount of work. And indeed, in practice, it really does need to do some work for each character.
>
> A Thompson NFA simulation needs to keep track of every state it is simultaneously in at any given point. And in order to compute the transition function, you need to compute it for every state you're in. The epsilon transitions that are added as part of the Thompson NFA construction (and are, crucially, what make building a Thompson NFA so fast) exacerbate this. So what happens is that you wind up chasing epsilon transitions over and over for each character.
>
> A DFA pre-computes these epsilon closures during powerset construction. Of course, that takes worst case O(2^m) time, which is why real DFAs aren't really used in general purpose engines. Instead, lazy DFAs are used.
>
> As for things like V8, they are backtrackers. They don't need to keep track of every state they're simultaneously in because they don't mind taking a very long time to complete some searches. But in practice, this can make them much faster for some inputs.
>
> Feel free to ask more questions. I'll stop here for now.

I also very briefly touch on this in my [blog post on regex internals](https://blog.burntsushi.net/regex-internals/#engine-pike-vm):

> The main problem with the PikeVM is performance, and its poor performance is primarily rooted in having to keep track of so much state. The capture group positions are required to report the start and end match offsets, while the currently active states must be tracked in order to guarantee worst case O(m * n) time. That is, in contrast to a backtracking approach, the PikeVM visits each byte in the haystack at most a constant number of times, and it does so by computing all possible active states in lock-step. This adds quite a bit of overhead, and it can be exacerbated by regexes with a lot of epsilon transitions. (This is one of the reasons why, earlier in this blog, we spent so much time talking about optimizations in the NFA to elide epsilon transitions.)

You can even see this in the code itself. For example, this bit is executed for every byte in a haystack:

https://github.com/rust-lang/regex/blob/837fd85e79fac2a4ea64030411b9a4a7b17dfa42/regex-automata/src/nfa/thompson/pikevm.rs#L1457-L1490

Compare that with a DFA, where the number of instructions executed per byte of haystack is constant and very small:

https://github.com/rust-lang/regex/blob/837fd85e79fac2a4ea64030411b9a4a7b17dfa42/regex-automata/src/dfa/search.rs#L97-L124
